blscurve/bls_batch_verifier.nim

```diff
@@ -0,0 +1,372 @@
+  ## Requires OpenMP 3.0 (GCC 4.4, 2008)
+  let numSets = batcher.sets.len.uint32
+  if numSets == 0:
+    # Spec precondition
+    return false
+
+  let numBatches = min(numSets, omp_get_max_threads().uint32)
+
+  # Stage 0: Accumulators - setLen for noinit of seq
+  batcher.batchContexts.setLen(numBatches.int)
+  batcher.updateResults.setLen(numBatches.int)
+
+  # No stacktrace, exception
+  # or anything that require a GC in a parallel section
+  # otherwise "attachGC()" is needed in the parallel prologue
+  # Hence we use raw ptr UncheckedArray instead of seq
+  let contextsPtr = batcher.batchContexts.toPtrUncheckedArray()
+  let setsPtr = batcher.sets.toPtrUncheckedArray()
+  let updateResultsPtr = batcher.updateResults.toPtrUncheckedArray()
+
+  # Stage 1: Accumulate partial pairings
+  checksAndStackTracesOff:
+    omp_parallel: # Start the parallel region
+      let threadID = omp_get_thread_num()
+      omp_chunks(numSets, chunkStart, chunkLen):
+        # Partition work into even chunks
+        # Each thread receives a different start+len to process
+        # chunkStart and chunkLen are set per-thread by the template
+        contextsPtr[threadID].init(
+          secureRandomBytes,
+          threadSepTag = cast[array[sizeof(threadID), byte]](threadID)
+        )
+
+        updateResultsPtr[threadID].ok =
+          accumPairingLines(
+            setsPtr, contextsPtr,
+            threadID.uint32,
+            chunkStart.uint32, uint32(chunkStart+chunkLen)
+          )
+
+  for i in 0 ..< batcher.updateResults.len:
+    if not updateResultsPtr[i].ok:
+      return false
+
+  # Stage 2: Reduce partial pairings
+  if numBatches < 4: # linear merge
+    for i in 1 ..< numBatches:
+      let ok = contextsPtr[0].merge(contextsPtr[i])
+      if not ok:
+        return false
+
+  else: # parallel logarithmic merge
+    var ok = false
+    checksAndStackTracesOff:
+      omp_parallel: # Start the parallel region
+        omp_single: # A single thread should create the master task
+          ok = reducePartialPairings(contextsPtr, start = 0, stopEx = numBatches)
+
+    if not ok:
+      return false
+
+  return batcher.batchContexts[0].finalVerify()
+
+# Autoselect Batch Verifier
+# ----------------------------------------------------------------------
+
+proc batchVerify*(
+       batcher: var BatchedBLSVerifier,
+       secureRandomBytes: array[32, byte]
+     ): bool =
+  ## Verify all signatures in batch at once.
+  ## Returns true if all signatures are correct
+  ## Returns false if there is at least one incorrect signature
+  ##
+  ## This requires securely generated random bytes
+  ## for scalar blinding
+  ## to defend against forged signatures that would not
+  ## verify individually but would verify while aggregated.
+  ##
+  ## The blinding scheme also assumes that the attacker cannot
+  ## resubmit 2^64 times forged (publickey, message, signature) triplets
+  ## against the same `secureRandomBytes`
+  if batcher.sets.len >= 3:
```

Comment:
Is this threshold appropriate for both few (2 or 4 cores, maybe) and many cores (16, 32, or 64 cores)?

Reply:
Yes, tested on a Pi with 4 cores and my 18 cores CPU.

Basically the synchronization operation needed are:

- 1 pairing commit per core (Miller Loop, 40% of single core pairing)
- 1 pairing merge per core (FP12 multiplication, 0.3% of a single core pairing)
- 1 pairing finalVerify (Final Exponentiation, 60 % of a single core pairing)

Batching, even serial saves the finalVerify hence why we have x2 acceleration with batching even of very small sets.

Now due to the tree algorithm, if we try to distribute only 2 sig verifications we would have
- T0: Sig0 -> commit -> sync -> merge -> finalVerify
- T1: Sig1 -> commit -> sync

Instead of
- T0: Sig0+Sig1 -> commit -> finalVerify

Sig0+Sig1 is cheaper than merge

It's possible that it should be 4 as well but that microtuning can be done later.